File: ts/earthbucks-lib/test/tx.test.ts
import { describe, expect, test, beforeEach, it } from "vitest";
import { Tx, HashCache } from "../src/tx.js";
import { TxIn } from "../src/tx-in.js";
import { TxOut } from "../src/tx-out.js";
import { Script } from "../src/script.js";
import { BufReader } from "../src/buf-reader.js";
import { BufWriter } from "../src/buf-writer.js";
import { Hash } from "../src/hash.js";
import { TxSignature } from "../src/tx-signature.js";
import { KeyPair } from "../src/key-pair.js";
import { FixedBuf, SysBuf } from "../src/buf.js";
import { U8, U16, U32, U64 } from "../src/numbers.js";
import { ScriptChunk } from "../src/script-chunk.js";
import { PrivKey } from "../src/priv-key.js";

describe("Tx", () => {
	describe("constructor", () => {
		test("should create a Tx", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [];
			const outputs: TxOut[] = [];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);
			expect(tx.version).toBe(version);
			expect(tx.inputs).toBe(inputs);
			expect(tx.outputs).toBe(outputs);
			expect(tx.lockAbs).toBe(lockAbs);
		});
	});

	test("to/from u8Vec", () => {
		const version = new U8(0);
		const inputs: TxIn[] = [
			new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
		];
		const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
		const lockAbs = new U32(0);

		const tx = new Tx(version, inputs, outputs, lockAbs);
		const result = Tx.fromBuf(tx.toBuf());
		expect(tx.toBuf().toString("hex")).toEqual(result.toBuf().toString("hex"));
	});

	describe("fromU8Vec", () => {
		test("fromU8Vec", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const result = Tx.fromBuf(tx.toBuf());
			expect(result.version).toEqual(version);
			expect(result.inputs.length).toEqual(inputs.length);
			expect(result.outputs.length).toEqual(outputs.length);
			expect(result.lockAbs).toEqual(lockAbs);
		});
	});

	describe("fromBufReader", () => {
		test("fromBufReader", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const reader = new BufReader(tx.toBuf());
			const result = Tx.fromBufReader(reader);
			expect(result.version).toEqual(version);
			expect(result.inputs.length).toEqual(inputs.length);
			expect(result.outputs.length).toEqual(outputs.length);
			expect(result.lockAbs).toEqual(lockAbs);
		});
	});

	describe("to/from string", () => {
		test("to/from string", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const result = Tx.fromHex(tx.toHex());
			expect(result.version).toEqual(version);
			expect(result.inputs.length).toEqual(inputs.length);
			expect(result.outputs.length).toEqual(outputs.length);
			expect(result.lockAbs).toEqual(lockAbs);
		});
	});

	describe("fromMintTx", () => {
		test("fromMintTx", () => {
			const script = Script.fromString("DOUBLEBLAKE3");
			const txInput = TxIn.fromMintTxScript(script);
			expect(txInput.inputTxId.buf.every((byte) => byte === 0)).toBe(true);
			expect(txInput.inputTxNOut.n).toEqual(0xffffffff);
			expect(txInput.script.toString()).toEqual(script.toString());
			expect(txInput.lockRel.n).toBe(0);
		});
	});

	describe("isMintTx", () => {
		test("isMintTx", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(
					FixedBuf.alloc(32),
					new U32(0xffffffff),
					new Script([
						ScriptChunk.fromData(SysBuf.alloc(32)),
						ScriptChunk.fromData(SysBuf.alloc(32)),
						ScriptChunk.fromData(SysBuf.from("example.com", "utf8")),
					]),
					new U32(0),
				),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);
			expect(tx.isMintTx()).toBe(true);
		});

		test("is not mintTx", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);
			expect(tx.isMintTx()).toBe(false);
		});

		test("fromMintTx -> isMintTx", () => {
			const script = new Script([
				ScriptChunk.fromData(SysBuf.alloc(32)),
				ScriptChunk.fromData(SysBuf.alloc(32)),
				ScriptChunk.fromData(SysBuf.from("example.com", "utf8")),
			]);
			const txInput = TxIn.fromMintTxScript(script);
			const tx = new Tx(new U8(0), [txInput], [], new U32(0));
			expect(tx.isMintTx()).toBe(true);
		});
	});

	describe("hashonce", () => {
		it("should return the hash of the tx", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);
			const expectedHash = Hash.blake3Hash(tx.toBuf());
			expect(tx.blake3Hash()).toEqual(expectedHash);
		});
	});

	describe("hash", () => {
		it("should return the hash of the hash of the tx", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);
			const expectedHash = Hash.blake3Hash(Hash.blake3Hash(tx.toBuf()).buf);
			expect(tx.id()).toEqual(expectedHash);
		});
	});

	describe("clone", () => {
		it("should verify this known tx clone", () => {
			const hexTx =
				"000100000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000640000000000";
			const tx = Tx.fromHex(hexTx);
			const clone = tx.clone();
			expect(tx.toHex()).toEqual(clone.toHex());
		});
	});

	describe("sighash", () => {
		test("hashPrevouts", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const result = tx.hashPrevouts();

			expect(SysBuf.from(result.buf).toString("hex")).toEqual(
				"2cb9ad7c6db72bb07dae3873c8a28903510eb87fae097338bc058612af388fba",
			);
		});

		test("hashLockRel", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const result = tx.hashLockRel();

			expect(SysBuf.from(result.buf).toString("hex")).toEqual(
				"406986f514581cacbf3ab0fc3863b336d137af79318ce4bae553a91435773931",
			);
		});

		test("hashOutputs", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(FixedBuf.alloc(32), new U32(0), new Script(), new U32(0)),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), new Script())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const result = tx.hashOutputs();

			expect(SysBuf.from(result.buf).toString("hex")).toEqual(
				"8c92e84e8b3b8b44690cbf64547018defaf43ade3b793ed8aa8ad33ae33941e5",
			);
		});

		test("sighash", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(
					FixedBuf.alloc(32),
					new U32(0),
					Script.fromEmpty(),
					new U32(0),
				),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), Script.fromEmpty())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const script = Script.fromEmpty();
			const scriptU8Vec = script.toBuf();
			const result = tx.sighashNoCache(
				new U32(0),
				scriptU8Vec,
				new U64(1),
				TxSignature.SIGHASH_ALL,
			);

			expect(result.toHex()).toEqual(
				"89a29cd97676864a55fc69ae7b3db3e220736e4ebda430d60985f48ee4e4b6af",
			);
		});

		test("sighash with cache", () => {
			const version = new U8(0);
			const inputs: TxIn[] = [
				new TxIn(
					FixedBuf.alloc(32),
					new U32(0),
					Script.fromEmpty(),
					new U32(0),
				),
			];
			const outputs: TxOut[] = [new TxOut(new U64(100), Script.fromEmpty())];
			const lockAbs = new U32(0);

			const tx = new Tx(version, inputs, outputs, lockAbs);

			const script = Script.fromEmpty();
			const scriptU8Vec = script.toBuf();
			const hashCache = new HashCache();
			const result = tx.sighashWithCache(
				new U32(0),
				scriptU8Vec,
				new U64(1),
				TxSignature.SIGHASH_ALL,
				hashCache,
			);

			expect(result.toHex()).toEqual(
				"89a29cd97676864a55fc69ae7b3db3e220736e4ebda430d60985f48ee4e4b6af",
			);
		});

		describe("sign and verify", () => {
			it("should generate a deterministic signature", () => {
				// Arrange
				const inputIndex = new U32(0);
				const privateKey = SysBuf.from(
					"7ca2df5597b60403be38cdbd4dc4cd89d7d00fce6b0773ef903bc8b87c377fad",
					"hex",
				);
				const script = SysBuf.from([]);
				const amount = new U64(100);
				const hashType = TxSignature.SIGHASH_ALL;
				const inputs: TxIn[] = [
					new TxIn(
						FixedBuf.alloc(32),
						new U32(0),
						Script.fromEmpty(),
						new U32(0),
					),
				];
				const outputs: TxOut[] = [new TxOut(new U64(100), Script.fromEmpty())];
				const tx = new Tx(new U8(0), inputs, outputs, new U32(0));

				// Act
				const signature = tx.signNoCache(
					inputIndex,
					PrivKey.fromBuf(FixedBuf.fromBuf(32, privateKey)),
					script,
					amount,
					hashType,
				);

				// Assert
				const expectedSignatureHex =
					"010664323b0876e7076e817efc29c3b7ffa0bab7de4fdb4a314601f9d9235a8a2202c2887491ee5a6c22a8fdbe554722794e92063cc47b5c7c711132dd357d08ad"; // your expected signature in hex
				expect(SysBuf.from(signature.toBuf()).toString("hex")).toEqual(
					expectedSignatureHex,
				);
			});

			it("should verify a deterministic signature", () => {
				// Arrange
				const inputIndex = new U32(0);
				const privateKey = FixedBuf.fromHex(
					32,
					"7ca2df5597b60403be38cdbd4dc4cd89d7d00fce6b0773ef903bc8b87c377fad",
				);
				const script = SysBuf.from([]);
				const amount = new U64(100);
				const hashType = TxSignature.SIGHASH_ALL;
				const inputs: TxIn[] = [
					new TxIn(
						FixedBuf.alloc(32),
						new U32(0),
						Script.fromEmpty(),
						new U32(0),
					),
				];
				// expect tx output to equal hext
				expect(inputs[0]?.toBuf().toString("hex")).toEqual(
					"0000000000000000000000000000000000000000000000000000000000000000000000000000000000",
				);
				const outputs: TxOut[] = [new TxOut(new U64(100), Script.fromEmpty())];
				expect(outputs[0]?.toBuf().toString("hex")).toEqual(
					"000000000000006400",
				);
				const tx = new Tx(new U8(0), inputs, outputs, new U32(0));
				expect(tx.toBuf().toString("hex")).toEqual(
					"000100000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000640000000000",
				);

				// Act
				const signature = tx.signNoCache(
					inputIndex,
					PrivKey.fromBuf(privateKey),
					script,
					amount,
					hashType,
				);

				// Assert
				const expectedSignatureHex =
					"010664323b0876e7076e817efc29c3b7ffa0bab7de4fdb4a314601f9d9235a8a2202c2887491ee5a6c22a8fdbe554722794e92063cc47b5c7c711132dd357d08ad"; // your expected signature in hex
				expect(SysBuf.from(signature.toBuf()).toString("hex")).toEqual(
					expectedSignatureHex,
				);
				const publicKey = KeyPair.fromPrivKeyEbxBuf(privateKey).pubKey.toBuf();
				const result = tx.verifyNoCache(
					inputIndex,
					publicKey.buf,
					signature,
					script,
					amount,
				);
				expect(result).toBe(true);
			});

			it("should verify a deterministic signature with hash cache", () => {
				// Arrange
				const inputIndex = new U32(0);
				const privateKey = FixedBuf.fromHex(
					32,
					"7ca2df5597b60403be38cdbd4dc4cd89d7d00fce6b0773ef903bc8b87c377fad",
				);
				const script = SysBuf.from([]);
				const amount = new U64(100);
				const hashType = TxSignature.SIGHASH_ALL;
				const inputs: TxIn[] = [
					new TxIn(
						FixedBuf.alloc(32),
						new U32(0),
						Script.fromEmpty(),
						new U32(0),
					),
				];
				// expect tx output to equal hext
				expect(inputs[0]?.toBuf().toString("hex")).toEqual(
					"0000000000000000000000000000000000000000000000000000000000000000000000000000000000",
				);
				const outputs: TxOut[] = [new TxOut(new U64(100), Script.fromEmpty())];
				expect(outputs[0]?.toBuf().toString("hex")).toEqual(
					"000000000000006400",
				);
				const tx = new Tx(new U8(0), inputs, outputs, new U32(0));
				expect(tx.toBuf().toString("hex")).toEqual(
					"000100000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000640000000000",
				);
				const hashCache1 = new HashCache();

				// Act
				const signature = tx.signWithCache(
					inputIndex,
					PrivKey.fromBuf(privateKey),
					script,
					amount,
					hashType,
					hashCache1,
				);

				// Assert
				const expectedSignatureHex =
					"010664323b0876e7076e817efc29c3b7ffa0bab7de4fdb4a314601f9d9235a8a2202c2887491ee5a6c22a8fdbe554722794e92063cc47b5c7c711132dd357d08ad"; // your expected signature in hex
				expect(SysBuf.from(signature.toBuf()).toString("hex")).toEqual(
					expectedSignatureHex,
				);
				const publicKey = KeyPair.fromPrivKeyEbxBuf(privateKey).pubKey.toBuf();
				const hashCache2 = new HashCache();
				const result = tx.verifyWithCache(
					inputIndex,
					publicKey.buf,
					signature,
					script,
					amount,
					hashCache2,
				);
				expect(result).toBe(true);
			});
		});
	});
});

File: rs/earthbucks_lib/tests/tx_test.rs
use earthbucks_lib::hash::*;
use earthbucks_lib::iso_buf_reader::*;
use earthbucks_lib::priv_key::PrivKey;
use earthbucks_lib::script::Script;
use earthbucks_lib::tx::*;
use earthbucks_lib::tx_in::*;
use earthbucks_lib::tx_out::*;
use earthbucks_lib::tx_signature::*;

#[test]
fn test_tx() -> Result<(), String> {
    let input_tx_id = vec![0; 32];
    let input_tx_index = 0;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let lock_rel = 0;
    let tx_input = TxIn::new(input_tx_id, input_tx_index, script, lock_rel);

    let value = 100;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let tx_output = TxOut::new(value, script);

    let version = 1;
    let inputs = vec![tx_input];
    let outputs = vec![tx_output];
    let lock_num = 0;
    let tx = Tx::new(version, inputs, outputs, lock_num);

    let buf = tx.to_iso_buf();
    let tx2 = Tx::from_iso_buf(buf).unwrap();
    assert_eq!(tx.version, tx2.version);
    assert_eq!(tx.inputs.len(), tx2.inputs.len());
    assert_eq!(tx.outputs.len(), tx2.outputs.len());
    assert_eq!(tx.lock_abs, tx2.lock_abs);
    Ok(())
}

#[test]
fn test_from_iso_buf_reader() -> Result<(), String> {
    let input_tx_id = vec![0; 32];
    let input_tx_index = 0;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let lock_rel = 0;
    let tx_input = TxIn::new(input_tx_id, input_tx_index, script, lock_rel);

    let value = 100;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let tx_output = TxOut::new(value, script);

    let version = 1;
    let inputs = vec![tx_input];
    let outputs = vec![tx_output];
    let lock_num = 0;
    let tx = Tx::new(version, inputs, outputs, lock_num);

    let buf = tx.to_iso_buf();
    let mut reader = IsoBufReader::new(buf);
    let tx2 = Tx::from_iso_buf_reader(&mut reader).unwrap();
    assert_eq!(tx.version, tx2.version);
    assert_eq!(tx.inputs.len(), tx2.inputs.len());
    assert_eq!(tx.outputs.len(), tx2.outputs.len());
    assert_eq!(tx.lock_abs, tx2.lock_abs);
    Ok(())
}

#[test]
fn test_to_string_and_from_iso_str() -> Result<(), String> {
    let input_tx_id = vec![0; 32];
    let input_tx_index = 0;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let lock_rel = 0;
    let tx_input = TxIn::new(input_tx_id, input_tx_index, script, lock_rel);

    let value = 100;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let tx_output = TxOut::new(value, script);

    let version = 1;
    let inputs = vec![tx_input];
    let outputs = vec![tx_output];
    let lock_num = 0;
    let tx = Tx::new(version, inputs, outputs, lock_num);

    let hex = tx.to_iso_str();
    let tx2 = Tx::from_iso_str(&hex).unwrap();
    assert_eq!(tx.version, tx2.version);
    assert_eq!(tx.inputs.len(), tx2.inputs.len());
    assert_eq!(tx.outputs.len(), tx2.outputs.len());
    assert_eq!(tx.lock_abs, tx2.lock_abs);
    Ok(())
}

#[test]
fn test_from_coinbase() {
    let input_script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let output_script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let output_amount = 100;
    let tx = Tx::from_coinbase(input_script, output_script, output_amount, 0);
    assert_eq!(tx.version, 1);
    assert_eq!(tx.inputs.len(), 1);
    assert_eq!(tx.outputs.len(), 1);
    assert_eq!(tx.lock_abs, 0);
}

#[test]
fn test_is_coinbase() {
    let input_tx_id = vec![0; 32];
    let input_tx_index = 0;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let lock_rel = 0;
    let tx_input = TxIn::new(input_tx_id, input_tx_index, script, lock_rel);

    let value = 100;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let tx_output = TxOut::new(value, script);

    let version = 1;
    let inputs = vec![tx_input];
    let outputs = vec![tx_output];
    let lock_num = 0;
    let tx = Tx::new(version, inputs, outputs, lock_num);
    assert!(!tx.is_coinbase());

    let input_script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let output_script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let output_amount = 100;
    let tx = Tx::from_coinbase(input_script, output_script, output_amount, 0);
    assert!(tx.is_coinbase());
}

#[test]
fn test_hash_once() {
    let input_tx_id = vec![0; 32];
    let input_tx_index = 0;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let lock_rel = 0;
    let tx_input = TxIn::new(input_tx_id, input_tx_index, script, lock_rel);

    let value = 100;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let tx_output = TxOut::new(value, script);

    let version = 1;
    let inputs = vec![tx_input];
    let outputs = vec![tx_output];
    let lock_num = 0;
    let tx = Tx::new(version, inputs, outputs, lock_num);
    let expected_hash = blake3_hash(&tx.to_iso_buf());
    assert_eq!(tx.blake3_hash(), expected_hash);
}

#[test]
fn test_hash() {
    let input_tx_id = vec![0; 32];
    let input_tx_index = 0;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let lock_rel = 0;
    let tx_input = TxIn::new(input_tx_id, input_tx_index, script, lock_rel);

    let value = 100;
    let script = Script::from_iso_str("DOUBLEBLAKE3 BLAKE3 DOUBLEBLAKE3 EQUAL").unwrap();
    let tx_output = TxOut::new(value, script);

    let version = 1;
    let inputs = vec![tx_input];
    let outputs = vec![tx_output];
    let lock_num = 0;
    let tx = Tx::new(version, inputs, outputs, lock_num);
    let expected_hash = double_blake3_hash(&tx.to_iso_buf());
    assert_eq!(tx.id(), expected_hash);
}

#[test]
fn test_hash_prevouts() {
    let version = 1;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    let outputs = vec![TxOut::new(100, Script::from_empty())];

    let tx = Tx::new(version, inputs, outputs, 0);

    let result = tx.hash_prevouts();

    assert_eq!(result.len(), 32);

    let expected =
        hex::decode("2cb9ad7c6db72bb07dae3873c8a28903510eb87fae097338bc058612af388fba").unwrap();
    assert_eq!(result, expected);
}

#[test]
fn test_hash_lock_rel() {
    let version = 1;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    let outputs = vec![TxOut::new(100, Script::from_empty())];

    let tx = Tx::new(version, inputs, outputs, 0);

    let result = tx.hash_lock_rel();

    assert_eq!(result.len(), 32);

    let expected =
        hex::decode("406986f514581cacbf3ab0fc3863b336d137af79318ce4bae553a91435773931").unwrap();
    assert_eq!(result, expected);
}

#[test]
fn test_hash_outputs() {
    let version = 1;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    let outputs = vec![TxOut::new(100, Script::from_empty())];

    let tx = Tx::new(version, inputs, outputs, 0);

    let result = tx.hash_outputs();

    assert_eq!(result.len(), 32);

    let expected =
        hex::decode("8c92e84e8b3b8b44690cbf64547018defaf43ade3b793ed8aa8ad33ae33941e5").unwrap();
    assert_eq!(result, expected);
}

#[test]
fn test_sighash() {
    let version = 1;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    let outputs = vec![TxOut::new(100, Script::from_empty())];

    let mut tx = Tx::new(version, inputs, outputs, 0);

    let script = Script::from_empty();
    let amount = 1;
    let hash_type = TxSignature::SIGHASH_ALL;
    let preimage = tx.sighash_no_cache(0, script.to_iso_buf(), amount, hash_type);

    let expected =
        hex::decode("a4f4519c65fedfaf43b7cc989f1bcdd55b802738d70f06ea359f411315b71c51").unwrap();
    assert_eq!(preimage, expected);
}

#[test]
fn test_sighash_with_cache() {
    let version = 1;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    let outputs = vec![TxOut::new(100, Script::from_empty())];

    let mut tx = Tx::new(version, inputs, outputs, 0);

    let script = Script::from_empty();
    let amount = 1;
    let hash_type = TxSignature::SIGHASH_ALL;
    let hash_cache = &mut HashCache::new();
    let preimage = tx.sighash_with_cache(0, script.to_iso_buf(), amount, hash_type, hash_cache);

    let expected =
        hex::decode("a4f4519c65fedfaf43b7cc989f1bcdd55b802738d70f06ea359f411315b71c51").unwrap();
    assert_eq!(preimage, expected);
}

#[test]
fn sign_and_verify() {
    // Arrange
    let input_index = 0;
    let private_key =
        hex::decode("7ca2df5597b60403be38cdbd4dc4cd89d7d00fce6b0773ef903bc8b87c377fad").unwrap();
    let script = vec![];
    let amount = 100;
    let hash_type = TxSignature::SIGHASH_ALL;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    assert_eq!(
        hex::encode(inputs[0].to_iso_buf()),
        "0000000000000000000000000000000000000000000000000000000000000000000000000000000000"
    );
    let outputs = vec![TxOut::new(100, Script::from_empty())];
    assert_eq!(hex::encode(outputs[0].to_iso_buf()), "000000000000006400");
    let mut tx = Tx::new(1, inputs, outputs, 0);
    assert_eq!(hex::encode(tx.to_iso_buf()), "01010000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000064000000000000000000");

    // Act
    let signature = tx.sign_no_cache(
        input_index,
        private_key.as_slice().try_into().unwrap(),
        script.clone(),
        amount,
        hash_type,
    );

    // Assert
    let expected_signature_hex = "0125c1e7312e2811c13952ea01e39f186cbc3077bef710ef11a363f88eae64ef0c657a1a6fd4bb488b69485f1ce7513fb3bab3cad418bc4f5093f648572f7fc89d";
    assert_eq!(hex::encode(signature.to_iso_buf()), expected_signature_hex);

    // Arrange
    // let key = KeyPair::new(private_key);
    // let public_key = key.public_key();

    let priv_key = PrivKey::from_iso_buf(private_key).unwrap();
    let pub_key_buf = priv_key.to_pub_key_buffer().unwrap();

    // Act
    let result = tx.verify_no_cache(input_index, pub_key_buf, signature, script.clone(), amount);

    // Assert
    assert!(result);
}

#[test]
fn sign_and_verify_with_cache() {
    // Arrange
    let input_index = 0;
    let private_key =
        hex::decode("7ca2df5597b60403be38cdbd4dc4cd89d7d00fce6b0773ef903bc8b87c377fad").unwrap();
    let script = vec![];
    let amount = 100;
    let hash_type = TxSignature::SIGHASH_ALL;
    let inputs = vec![TxIn::new(vec![0; 32], 0, Script::from_empty(), 0)];
    assert_eq!(
        hex::encode(inputs[0].to_iso_buf()),
        "0000000000000000000000000000000000000000000000000000000000000000000000000000000000"
    );
    let outputs = vec![TxOut::new(100, Script::from_empty())];
    assert_eq!(hex::encode(outputs[0].to_iso_buf()), "000000000000006400");
    let mut tx = Tx::new(1, inputs, outputs, 0);
    assert_eq!(hex::encode(tx.to_iso_buf()), "01010000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000064000000000000000000");
    let hash_cache_1 = &mut HashCache::new();

    // Act
    let signature = tx.sign_with_cache(
        input_index,
        private_key.as_slice().try_into().unwrap(),
        script.clone(),
        amount,
        hash_type,
        hash_cache_1,
    );

    // Assert
    let expected_signature_hex = "0125c1e7312e2811c13952ea01e39f186cbc3077bef710ef11a363f88eae64ef0c657a1a6fd4bb488b69485f1ce7513fb3bab3cad418bc4f5093f648572f7fc89d";
    assert_eq!(hex::encode(signature.to_iso_buf()), expected_signature_hex);

    // Arrange
    // let key = KeyPair::new(private_key);
    // let public_key = key.public_key();
    let hash_cache_2 = &mut HashCache::new();
    let pub_key_buf = PrivKey::from_iso_buf(private_key)
        .unwrap()
        .to_pub_key_buffer()
        .unwrap();

    // Act
    let result = tx.verify_with_cache(
        input_index,
        pub_key_buf,
        signature,
        script.clone(),
        amount,
        hash_cache_2,
    );

    // Assert
    assert!(result);
}
